### ISSUE TITLE
examples/dwarf_lineprogram_filenames.py: TypeError: tuple indices must be integers or slices, not str

### ISSUE THREAD

**Issue by vegard** [eliben/pyelftools#417]
I cloned a fresh copy of pyelftools to check DWARF5 line program support and I get this exception:

```
pyelftools$ python3 examples/dwarf_lineprogram_filenames.py --test /tmp/tmp31cx7ko3/usr/lib/debug/lib/modules/5.17.0-3-amd64/kernel/drivers/atm/ambassador.ko
Processing file: /tmp/tmp31cx7ko3/usr/lib/debug/lib/modules/5.17.0-3-amd64/kernel/drivers/atm/ambassador.ko
  Found a compile unit at offset 0, length 147144
Traceback (most recent call last):
  File "examples/dwarf_lineprogram_filenames.py", line 96, in <module>
    process_file(filename)
  File "examples/dwarf_lineprogram_filenames.py", line 46, in process_file
    line_entry_mapping(line_program)
  File "examples/dwarf_lineprogram_filenames.py", line 62, in line_entry_mapping
    filename = lpe_filename(line_program, lpe.state.file)
  File "examples/dwarf_lineprogram_filenames.py", line 82, in lpe_filename
    dir_index = file_entry["dir_index"]
TypeError: tuple indices must be integers or slices, not str
```

This is the unmodified example program that comes with pyelftools.

**Comment by sevaa** [comment]
The first one I sort of know what to do (I broke it in the first place :)). The second one will need a more careful look. I'm on another piece at the moment, but I'll get to this. Thanks for bringing it up.

**Comment by vegard** [comment]
```
$ python3 --version
Python 3.8.10
```

**Comment by vegard** [comment]
Nice, if I just change it like this, it works:
```
--- a/examples/dwarf_lineprogram_filenames.py
+++ b/examples/dwarf_lineprogram_filenames.py
@@ -79,7 +79,7 @@ def lpe_filename(line_program, file_index):
 
     # File and directory indices are 1-indexed.
     file_entry = file_entries[file_index - 1]
-    dir_index = file_entry["dir_index"]
+    dir_index = file_entry.dir_index
 
     # A dir_index of 0 indicates that no absolute directory was recorded during
     # compilation; return just the basename.
```
output:
```
$ python3 examples/dwarf_lineprogram_filenames.py --test /tmp/tmp31cx7ko3/usr/lib/debug/lib/modules/5.17.0-3-amd64/kernel/drivers/atm/ambassador.ko
Processing file: /tmp/tmp31cx7ko3/usr/lib/debug/lib/modules/5.17.0-3-amd64/kernel/drivers/atm/ambassador.ko
  Found a compile unit at offset 0, length 147144
    filename=debian/build/build_amd64_none_amd64/ambassador.c -> 290 entries
    filename=drivers/atm/ihex.h -> 2607 entries
    filename=include/linux/io.h -> 162 entries
    filename=drivers/atm/spinlock.h -> 12 entries
    filename=include/asm-generic/bitops/atomic-instrumented.h -> 12 entries
    filename=drivers/atm/device.h -> 2 entries
    filename=include/linux/atomic/little_endian.h -> 4 entries
    filename=include/linux/bitops.h -> 51 entries
    filename=include/uapi/linux/byteorder/swab.h -> 23 entries
    filename=include/linux/atomic.h -> 62 entries
    filename=include/linux/page_64.h -> 66 entries
    filename=arch/x86/include/asm/instrumented-non-atomic.h -> 106 entries
    filename=drivers/atm/fortify-string.h -> 4 entries
    filename=drivers/atm/interrupt.h -> 8 entries
    filename=drivers/atm/slab.h -> 93 entries
    filename=include/linux/current.h -> 9 entries
    filename=drivers/atm/skbuff.h -> 16 entries
    filename=drivers/atm/netdevice.h -> 3 entries
    filename=drivers/atm/timekeeping.h -> 14 entries
    filename=arch/x86/include/asm/instrumented-atomic.h -> 4 entries
    filename=drivers/atm/pci.h -> 8 entries
  Found a compile unit at offset 147148, length 26074
```

**Comment by sevaa** [comment]
It's probably not the only consumer of pyelftools that relies on file_entries being indexable by field name. file_entries is generated differently for DWARF5, I should mimic the pre-DWARF5 behavior.

**Comment by sevaa** [comment]
For a fix for the "TypeError: tuple indices must be integers or slices, not str" error, repeat the following commit on your side: https://github.com/eliben/pyelftools/commit/80f6a852cf32ea9affa8c31fbb92d127d5fda09c

**Comment by vegard** [comment]
Maybe this should be a separate issue, but I think there's maybe something off about the directory name computations in the DWARF5 support:

```
$ python3 examples/dwarf_lineprogram_filenames.py --test drivers/acpi/acpi_pad.ko
Processing file: drivers/acpi/acpi_pad.ko
  Found a compile unit at offset 0, length 55301
    filename=debian/build/build_amd64_none_amd64/acpi_pad.c -> 569 entries
[...]
```

`debian/build/build_amd64_none_amd64/acpi_pad.c` is not the correct path for this file, you can tell because gdb recognizes the correct path:

```
$ gdb drivers/acpi/acpi_pad.ko
[...]
(gdb) break power_saving_thread
Breakpoint 1 at 0x660: file drivers/acpi/acpi_pad.c, line 139.
```

zipped `acpi_pad.ko`: [acpi_pad.zip](https://github.com/eliben/pyelftools/files/8903793/acpi_pad.zip)

**Comment by vegard** [comment]
I can fix the previous issue like this:
```
--- a/examples/dwarf_lineprogram_filenames.py
+++ b/examples/dwarf_lineprogram_filenames.py
@@ -86,7 +86,7 @@ def lpe_filename(line_program, file_index):
     if dir_index == 0:
         return file_entry.name.decode()
 
-    directory = lp_header["include_directory"][dir_index - 1]
+    directory = lp_header["include_directory"][dir_index]
     return posixpath.join(directory, file_entry.name).decode()
 
 
```

then I get:
```
$ python3 examples/dwarf_lineprogram_filenames.py --test drivers/acpi/acpi_pad.ko
Processing file: drivers/acpi/acpi_pad.ko
  Found a compile unit at offset 0, length 55301
    filename=drivers/acpi/acpi_pad.c -> 569 entries
```

I'm not sure if this is correct for DWARF4 though.

**Comment by sevaa** [comment]
I'll check the spec.

In readelf.py, method `_dump_debug_line_programs()`, there's unconditional logic: dir_index 0 means current, dir_index>0 means `include_directory[dir_index-1]`. Just saying'.

**Comment by sevaa** [comment]
Looks like in DWARFv5 lineprograms, they've abandoned the "index 0 means default, index 1 or higher means index-1" logic for files and directories.

From DWARFv4, section 6.2.4, 11/12

> The line number program assigns numbers to each of the file entries in order, **beginning with 1**.
> The line number program assigns numbers to each of the file entries in order, **beginning with 1**, and uses those numbers instead of file names in the file register.

The corresponding bit in DWARFv5, section 6.2.4, 16/20

> The line number program assigns a number (index) to each of the directory entries in order, **beginning with 0**.
> The line number program references file names in this sequence **beginning with 0**, and uses those numbers instead of file names in the line number program that follows.

There you have it. Explicitly changed in DWARF5.

**Comment by sevaa** [comment]
@vegard: this should be fixed in 0.29.

**Comment by eliben** [comment]
Closing; please reopen if you still experience this problem in the git master branch
